pkg/controllers/cassandra/pilot/pilot.go

```diff
@@ -0,0 +1,165 @@
+	allPods, err := c.pods.Pods(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return clusterPods, err
+	}
+	for _, pod := range allPods {
+		podControlledByCluster, err := controllers.PodControlledByCluster(
+			cluster,
+			pod,
+			c.statefulSets,
+		)
+		if err != nil {
+			return clusterPods, err
+		}
+		if !podControlledByCluster {
+			continue
+		}
+		clusterPods = append(clusterPods, pod)
+	}
+	return clusterPods, nil
+}
+
+func (c *pilotControl) createOrUpdatePilot(cluster *v1alpha1.CassandraCluster, pod *v1.Pod) error {
+	desiredPilot := PilotForCluster(cluster, pod)
+	client := c.naviClient.NavigatorV1alpha1().Pilots(desiredPilot.GetNamespace())
+	lister := c.pilots.Pilots(desiredPilot.GetNamespace())
+	existingPilot, err := lister.Get(desiredPilot.GetName())
+	if k8sErrors.IsNotFound(err) {
+		_, err = client.Create(desiredPilot)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	err = util.OwnerCheck(existingPilot, cluster)
+	if err != nil {
+		return err
+	}
+	desiredPilot = existingPilot.DeepCopy()
+	updatePilotForCluster(cluster, pod, desiredPilot)
+	_, err = client.Update(desiredPilot)
+	return err
+}
+
+func (c *pilotControl) removeUnusedPilots(
+	cluster *v1alpha1.CassandraCluster,
+) error {
+	expectedPilotNames := map[string]bool{}
+	clusterPods, err := c.clusterPods(cluster)
+	if err != nil {
+		return err
+	}
+	for _, pod := range clusterPods {
+		expectedPilotNames[pod.Name] = true
+	}
+	existingPilots, err := c.pilots.Pilots(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	client := c.naviClient.NavigatorV1alpha1().Pilots(cluster.Namespace)
+	for _, pilot := range existingPilots {
+		err := util.OwnerCheck(pilot, cluster)
+		if err != nil {
+			return err
+		}
+		_, found := expectedPilotNames[pilot.Name]
+		if !found {
+			err := client.Delete(pilot.Name, nil)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *pilotControl) syncPilots(cluster *v1alpha1.CassandraCluster) error {
+	pods, err := c.clusterPods(cluster)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods {
+		err = c.createOrUpdatePilot(cluster, pod)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (c *pilotControl) Sync(cluster *v1alpha1.CassandraCluster) error {
+	err := c.syncPilots(cluster)
+	if err != nil {
+		return err
+	}
+	err = c.removeUnusedPilots(cluster)
+	return err
+}
+
+func PilotForCluster(cluster *v1alpha1.CassandraCluster, pod *v1.Pod) *v1alpha1.Pilot {
+	pilot := &v1alpha1.Pilot{}
+	ownerRefs := pilot.GetOwnerReferences()
+	ownerRefs = append(ownerRefs, util.NewControllerRef(cluster))
+	pilot.SetOwnerReferences(ownerRefs)
+	return updatePilotForCluster(cluster, pod, pilot)
+}
+
+func updatePilotForCluster(
+	cluster *v1alpha1.CassandraCluster,
+	pod *v1.Pod,
+	pilot *v1alpha1.Pilot,
+) *v1alpha1.Pilot {
+	pilot.SetName(pod.GetName())
+	pilot.SetNamespace(cluster.GetNamespace())
+	pilot.SetLabels(util.ClusterLabels(cluster))
```

Comment:
We should be careful not to override any user provided labels, or labels that may have been added by the pilot (we don't do this right now, but we might do at some point)

Reply:
Done.